remote/gitea/gitea.go

```diff
@@ -0,0 +1,344 @@
+
+	client := c.newClientToken(u.Token)
+	all, err := client.ListMyRepos()
+	if err != nil {
+		return repos, err
+	}
+
+	for _, repo := range all {
+		repos = append(repos, toRepoLite(repo))
+	}
+	return repos, err
+}
+
+// Perm returns the user permissions for the named Gitea repository.
+func (c *client) Perm(u *model.User, owner, name string) (*model.Perm, error) {
+	client := c.newClientToken(u.Token)
+	repo, err := client.GetRepo(owner, name)
+	if err != nil {
+		return nil, err
+	}
+	return toPerm(repo.Permissions), nil
+}
+
+// File fetches the file from the Gitea repository and returns its contents.
+func (c *client) File(u *model.User, r *model.Repo, b *model.Build, f string) ([]byte, error) {
+	client := c.newClientToken(u.Token)
+	ref := b.Commit
+
+	// TODO gitea does not yet return a sha with the pull request
+	// so unfortunately we need to use the pull request branch.
+	if b.Event == model.EventPull {
+		ref = b.Branch
+	}
+	if ref == "" {
+		// Remove refs/tags or refs/heads, Gitea needs a short ref
+		ref = strings.TrimPrefix(
+			strings.TrimPrefix(
+				b.Ref,
+				"refs/heads/",
+			),
+			"refs/tags/",
+		)
+	}
+	cfg, err := client.GetFile(r.Owner, r.Name, ref, f)
+	return cfg, err
+}
+
+// FileRef fetches the file from the Gitea repository and returns its contents.
+func (c *client) FileRef(u *model.User, r *model.Repo, ref, f string) ([]byte, error) {
+	return c.newClientToken(u.Token).GetFile(r.Owner, r.Name, ref, f)
+}
+
+// Status is not supported by the Gitea driver.
+func (c *client) Status(u *model.User, r *model.Repo, b *model.Build, link string) error {
+	client := c.newClientToken(u.Token)
+
+	status := getStatus(b.Status)
+	desc := getDesc(b.Status)
+
+	_, err := client.CreateStatus(
+		r.Owner,
+		r.Name,
+		b.Commit,
+		gitea.CreateStatusOption{
+			status,
+			link,
+			desc,
+			"",
+		},
+	)
+
+	return err
+}
+
+// Netrc returns a netrc file capable of authenticating Gitea requests and
+// cloning Gitea repositories. The netrc will use the global machine account
+// when configured.
+func (c *client) Netrc(u *model.User, r *model.Repo) (*model.Netrc, error) {
+	if c.Password != "" {
+		return &model.Netrc{
+			Login:    c.Username,
+			Password: c.Password,
+			Machine:  c.Machine,
+		}, nil
+	}
+	return &model.Netrc{
+		Login:    u.Token,
+		Password: "x-oauth-basic",
+		Machine:  c.Machine,
+	}, nil
+}
+
+// Activate activates the repository by registering post-commit hooks with
+// the Gitea repository.
+func (c *client) Activate(u *model.User, r *model.Repo, link string) error {
+	config := map[string]string{
+		"url":          link,
+		"secret":       r.Hash,
+		"content_type": "json",
+	}
+	hook := gitea.CreateHookOption{
+		Type:   "gogs",
+		Config: config,
+		Events: []string{"push", "create", "pull_request"},
+		Active: true,
+	}
+
+	client := c.newClientToken(u.Token)
+	_, err := client.CreateRepoHook(r.Owner, r.Name, hook)
+	return err
+}
+
+// Deactivate is not supported by the Gitea driver.
```

Comment:
yes, we should implement Deactivate as part of this PR. Without the ability to deactivate, a repository cannot be re-enabled correctly